src/mango/src/mango_cursor_text.erl

```diff
@@ -334,4 +350,805 @@ get_json_docs(DbName, Hits) ->
+                limit = _L,
+                bookmark = B,
+                sort = relevance,
+                raw_bookmark = true
+            } = QueryArgs,
+            {Bookmark, Hits} =
+                case B of
+                    nil ->
+                        Hit1 = #sortable{item = #hit{fields = [{<<"_id">>, {id, 1}}]}},
+                        Hit2 = #sortable{item = #hit{fields = [{<<"_id">>, {id, 2}}]}},
+                        Hit3 = #sortable{item = #hit{fields = [{<<"_id">>, {id, 3}}]}},
+                        {[bookmark, 0], [Hit1, Hit2, Hit3]};
+                    [bookmark, 3] ->
+                        Hit1 = #sortable{item = #hit{fields = [{<<"_id">>, {id, 1}}]}},
+                        Hit2 = #sortable{item = #hit{fields = [{<<"_id">>, {id, 2}}]}},
+                        Hit3 = #sortable{item = #hit{fields = [{<<"_id">>, {id, 3}}]}},
+                        {[bookmark, 4], [Hit3, Hit2, Hit1]};
+                    [bookmark, N] ->
+                        {[bookmark, N + 1], []}
+                end,
+            {ok, Bookmark, undefined, Hits, undefined, undefined}
+        end
+    ),
+    meck:expect(mango_selector_text, convert, [selector], meck:val(query)),
+    meck:expect(mango_selector, match, fun(selector, {doc, _N}) -> true end),
+    ?assertEqual({ok, {acc, 6}}, execute(Cursor, fun foo:normal/2, {acc, 0})),
+    ?assertEqual(6, meck:num_calls(couch_stats, increment_counter, 1)),
+    ?assertEqual(6, meck:num_calls(mango_execution_stats, incr_docs_examined, 1)),
+    ?assertEqual(3, meck:num_calls(mango_execution_stats, incr_results_returned, 1)).
+
+t_execute_limit_cutoff(_) ->
+    Limit = 2,
+    Skip = 0,
+    IdxName = <<"index">>,
+    Sort = [{<<"field1">>, <<"desc">>}, {<<"field2">>, <<"asc">>}],
+    Options = [{partition, partition}, {sort, {Sort}}, {bookmark, []}],
+    Cursor = #cursor{
+        db = db,
+        index = #idx{ddoc = <<"_design/ddoc">>, name = IdxName},
+        limit = Limit,
+        skip = Skip,
+        fields = fields,
+        selector = selector,
+        opts = Options,
+        execution_stats = stats
+    },
+    QueryArgs =
+        #index_query_args{
+            q = query,
+            partition = partition,
+            limit = Skip + Limit,
+            bookmark = nil,
+            sort = [<<"-field1<type>">>, <<"field2<type>">>],
+            raw_bookmark = true
+        },
+    Hit1 = #sortable{item = #hit{fields = [{<<"_id">>, {id, 1}}]}},
+    Hit2 = #sortable{item = #hit{fields = [{<<"_id">>, {id, 2}}]}},
+    Hit3 = #sortable{item = #hit{fields = [{<<"_id">>, {id, 3}}]}},
+    Hits = [Hit1, Hit2, Hit3],
+    meck:expect(
+        dreyfus_fabric_search,
+        go,
+        [db_name, <<"ddoc">>, IdxName, QueryArgs],
+        meck:val({ok, [bookmark, 1], undefined, Hits, undefined, undefined})
+    ),
+    meck:expect(mango_selector_text, convert, [selector], meck:val(query)),
+    meck:expect(mango_selector, match, fun(selector, {doc, _N}) -> true end),
```

Comment:
If we care to highlight the variable name it might be better as `_Doc`. If we just care about `{doc, ...}` then `{doc, _}` might look better.  If we wanted to be more strict `{[{_, _} | _]}` might work but `_N` seems a bit odd on it's own as a variable name.

Reply:
It is only `{doc, _}` that matters.  I left the `N` suffix around because that is how it is used at the other mock function definitions.  `_` itself constitutes an anonymous variable, so extending it with a name and making it `_N` is more about documentation for the reader.

Mind that `{doc, N}` is not valid JSON document here (see my comment above on the reasoning) it just a token to check if the document variable was passed on properly.

Reply:
Sometimes it's worth naming un-used variables when it's not clear what the variables are, for instance:

```
myfun(UsedVar, _UnusedVar1, _UnusedVar2, _UnusedVar3, UsedVar2) -> ...
```

These will actually be bound and can be used further down, but that's not idiomatic. But the key is they'd be named for documentation purposes mostly, taking care we don't actually match anything with them or use them.

In case `mango_selector:match(Selector, Doc)` it does look a bit odd as we're matching a `doc` record but matching it as a tuple `{doc, Id, etc...}`. If that's the case we could just use `#doc{}` as a record here?

Reply:
Using a `{doc, N}` token is better to keep track of the proper data flow, for example.  In the mock definition of `mango_fields:extract/2` the relationship between the input and the output is described as a function that produces the final document and if that is not what is used at some later point then there is a glitch in the flow.  With regular values this becomes almost impossible to check, I believe.